### PR TITLE
Introduce `DTypePolicyMap`

### PR DESCRIPTION
**Pull request by james77777778**
I have tried to implement the idea mentioned by @mattdangerw :
https://github.com/keras-team/keras-nlp/pull/1612

I think this idea is clean and only requires some small changes to support mixing dtype policies into subclasses.

In general, `DTypePolicyMap` is similar to `LayoutMap` as it contains a dict to record the (layer name, policy) pair.
The limitations / overheads of using `DTypePolicyMap`:
- Each layer must have a unique name (`auto_name` might fail?)
- Some extra logic needs to be added in `get_config` for the top class (ex: `Backbone` in KerasCV and KerasNLP)

The basic usage:

<details>

```python
import keras
from keras import dtype_policies
from keras import layers


@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.Layer):
    def __init__(self, dtype=None, name="my_subclass", **kwargs):
        super().__init__(dtype=dtype, **kwargs)
        self.dense = layers.Dense(8, dtype=dtype, name=f"{name}_dense")
        self.bn = layers.BatchNormalization(dtype=dtype, name=f"{name}_bn")
        self.act = layers.Activation("relu", dtype=dtype, name=f"{name}_act")

    def call(self, inputs, training=None):
        return self.act(self.bn(self.dense(inputs), training=training))

    def get_config(self):
        config = super().get_config()
        # `default_policy=None` enables us to defer to
        # `keras.config.dtype_policy` during loading.
        dtype_policy_map = dtype_policies.DTypePolicyMap(default_policy=None)
        for layer in self._flatten_layers():
            if layer.dtype_policy.is_quantized:
                dtype_policy_map[layer.name] = layer.dtype_policy
        if len(dtype_policy_map) > 0:
            config.update({"dtype": dtype_policy_map})
        return config


inputs = keras.Input(shape=[None, 4])
outputs = MySubclass()(inputs)
model = keras.Model(inputs, outputs)

print(f"Saving stage, global policy: {keras.config.dtype_policy()}")
model.quantize("int8")
for layer in model._flatten_layers(include_self=False, recursive=True):
    print(layer.name, layer.dtype_policy)
model.save("model.keras")

keras.config.set_dtype_policy("bfloat16")
print(f"Loading stage, global policy: {keras.config.dtype_policy()}")
new_model = keras.models.load_model("model.keras")
for layer in new_model._flatten_layers(include_self=False, recursive=True):
    print(layer.name, layer.dtype_policy)

```

</details>

Outputs:
```bash
Saving stage, global policy: <FloatDTypePolicy "float32">
input_layer <FloatDTypePolicy "float32">
my_subclass <FloatDTypePolicy "float32">
my_subclass_act <FloatDTypePolicy "float32">
my_subclass_bn <FloatDTypePolicy "float32">
my_subclass_dense <QuantizedDTypePolicy "int8_from_float32">

Loading stage, global policy: <FloatDTypePolicy "bfloat16">
input_layer <FloatDTypePolicy "bfloat16">
my_subclass_1 <FloatDTypePolicy "bfloat16">
my_subclass_act <FloatDTypePolicy "bfloat16">
my_subclass_bn <FloatDTypePolicy "bfloat16">
my_subclass_dense <QuantizedDTypePolicy "int8_from_bfloat16">
```

### CODE REVIEW

File: keras/src/dtype_policies/dtype_policy_map.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import re
+from collections.abc import MutableMapping
+
+from keras.src import dtype_policies
+from keras.src.api_export import keras_export
+
+
+@keras_export(["keras.dtype_policies.DTypePolicyMap"])
+class DTypePolicyMap(MutableMapping):
```
Comment (mattdangerw) [inline keras/src/dtype_policies/dtype_policy_map.py:9]: Would we want this to be a DTypePolicy subclass? Or have dtype policy accept this as a sub-component? Might be nice to keep everything that is passed as a dtype eventually resolving to a `DTypePolicy` object.

File: keras/src/ops/operation.py [hunk 2]
```diff
@@ -23,7 +23,10 @@ def __init__(self, dtype=None, name=None):
                 "cannot contain character `/`. "
                 f"Received: name={name} (of type {type(name)})"
             )
-        self._dtype_policy = dtype_policies.get(dtype)
+        if isinstance(dtype, dtype_policies.DTypePolicyMap):
```
Comment (mattdangerw) [inline keras/src/ops/operation.py:26]: A disadvantage of doing this with operation name is it wouldn't resolve a whole path right?

[Here's how this works](https://github.com/keras-team/keras/blob/33756c12ea19a4a61cfeb1e284b29a109bce522e/keras/src/distribution/distribution_lib.py#L582-L587) for a `LayoutMap`. This uses `variable.path` to match a regex, so you can specify things based not just on a variable name or layer name, but a whole hierarchy. So `"encoder.key_dense.kernel"` get one value and `"encoder.value_dense.*` get's another.

Involving a path hierarchy would be more consistent with `LayoutMap`, but opens questions. How do you compute a path like this for a layer? I think it's just a variable concept right now.

Also, would we want this to allow setting different variable dtypes for different variables in the same layer? Are there real applications for doing that? And how practical would that be?

**Review by mattdangerw — COMMENTED** [review]
Left some quick questions, but @fchollet what's your thoughts on this as a whole?

Basically a way to specify a single dtype object that captures dtype information for a whole model, and can be directly passed layer to layer.

**Comment by james77777778** [comment]
Hey @mattdangerw 

If I'm not misunderstanding, `DTypePolicy` belongs to a layer's scope and we use it to parse the `compute_dtype` and `variable_dtype`. In the quantization, we also use `DTypePolicy` to determine whether to dispatch `call` or `quantized_call` in `Operation`.

So, IMO, `DTypePolicyMap` is a mapping between "layers" and "dtype polices", not "variables". The relationship is similar to `LayoutMap` which maps "variable.path" and "layout".

> How do you compute a path like this for a layer? I think it's just a variable concept right now.

This is an issue for `DTypePolicyMap`. That's one of the limitations mentioned in my first comment.
Things become challenging because we might have a different `name` for the exact same layer using `auto_name`.
(EX: instantiate the same model multiple times)

> Also, would we want this to allow setting different variable dtypes for different variables in the same layer? Are there real applications for doing that? And how practical would that be?

I don't think so.  I think it is rare or even non-existent to have scenarios where different dtypes are set for different variables within a layer.

**Comment by james77777778** [comment]
Hi @mattdangerw @fchollet 
Do we have any plans for this? This functionality could simplify supporting quantization for KerasNLP/KerasCV.

Or do we have another good way to achieve that?

**Comment by mattdangerw** [comment]
Sorry at @james77777778 was hoping @fchollet would take a look, totally agree we want something like this!

Let me try to ping him now.

**Comment by james77777778** [comment]
No hurry! Just curious how things are going.

**Comment by fchollet** [comment]
The overall approach and the usage code snippet look good to me. Keep a single object all dtype information (which would typically be programmatically generated from an existing quantized model), and pass it around to sublayers in the constructor. It's pretty clean. It's better than using scopes.

**Comment by james77777778** [comment]
Great! I will update this PR by adding docstrings and tests soon.

**Comment by james77777778** [comment]
Should be ready for reviewing.
I have added tests including a basic usage example.

File: keras/src/ops/operation.py [hunk 3]
```diff
@@ -23,7 +24,11 @@ def __init__(self, dtype=None, name=None):
...
+            self._dtype_policy = dtype_policies.get(dtype[name])
```
Comment (mattdangerw) [inline keras/src/ops/operation.py:28]: I think I am still missing something here. How is this supported to work for many layers of nesting? E.g. the KerasNLP case, where Model -> Block x 12 -> MHA x 2 -> Dense x 4.

In some cases, child layers are only created on `build()`, how do you proxy this object all the way down in that case? What the new best practice way to write a composite layer (potentially of more composite layers)?

And is this line still valid https://github.com/keras-team/keras/blob/v3.3.3/keras/src/layers/attention/multi_head_attention.py#L291

**Review by mattdangerw — COMMENTED** [review]
I might still not be grokking how this works, but I don't get how this is supposed to work for composite layers. Left a comment down below.

**Comment by james77777778** [comment]
@mattdangerw , @fchollet 
I think I have messed up the commit history. Sorry about that.
I have resummited the PR at: https://github.com/keras-team/keras/pull/19858

This PR should be closed now.

File: keras/src/dtype_policies/dtype_policy_map.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    """A dict-like object that maps string to `DTypePolicy` instances.
+
+    `DTypePolicyMap` uses a string as key and a `DTypePolicy` as value. There is
+    a behavior difference between a normal Python dict and this class. The
+    string key will be treated as a regex when retrieving the value. See the
+    docstring of `get` for more details.
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras/src/dtype_policies/dtype_policy_map.py]: I have updated the docstring in #19858.
This includes the reason why we need `DTypePolicyMap` and an example illustrating its use for adding quantization support to MHA.

**Pull request closed without merging**